lib/puppet/transaction/event_manager.rb

```diff
@@ -144,18 +144,33 @@ def process_callback(resource, callback, events)
       return false
     end
 
+    message = "Triggered '#{callback}' from #{events.length} events"
+    event = create_callback_status_event(resource, callback, message) if not resource.is_a?(Puppet::Type.type(:whit))
+
     resource.send(callback)
 
-    if not resource.is_a?(Puppet::Type.type(:whit))
-      resource.notice "Triggered '#{callback}' from #{events.length} events"
-    end
+    resource.notice message if not resource.is_a?(Puppet::Type.type(:whit))
     return true
   rescue => detail
-    resource.err "Failed to call #{callback}: #{detail}"
+    resource_error_message = "Failed to call #{callback}: #{detail}"
+    resource.err resource_error_message
+    if not resource.is_a?(Puppet::Type.type(:whit))
+      event = create_callback_status_event(resource, callback, resource_error_message)
```

Comment:
Could we simplify this to (so we don't overwrite the same `event` object)?
```ruby
  resource.send(callback)
  if not resource.is_a?(Puppet::Type.type(:whit))
    event =  resource.event(message: "Triggered '#{callback}' from #{events.length} events", status: "success", name: callback)
    transaction.resource_status(resource) << event
  end
  return true
rescue => detail
  resource_error_message = "Failed to call #{callback}: #{detail}"
  resource.err resource_error_message
  if not resource.is_a?(Puppet::Type.type(:whit))
    event =  resource.event(message: resource_error_message, status: "failed", name: callback)
    transaction.resource_status(resource) << event
  end
end
```

I imagine it's setup this way so that we can `ensure` an event is always written. But if an exception is raised which doesn't derive from `StandardError`, then we'll skip the `rescue` block, but log that the event was successful, which is false.